Subject: [PATCH] riscv: dts: sophgo: add SG2002 NanoKVM PCIe

diff --git a/arch/riscv/boot/dts/sophgo/Makefile b/arch/riscv/boot/dts/sophgo/Makefile
index 1ff69cf7f..797ff6c33 100644
--- a/arch/riscv/boot/dts/sophgo/Makefile
+++ b/arch/riscv/boot/dts/sophgo/Makefile
@@ -3,6 +3,7 @@ dtb-$(CONFIG_ARCH_SOPHGO) += cv1800b-milkv-duo.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += cv1812h-huashan-pi.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2002-licheerv-nano-b.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2002-milkv-duo256m.dtb
+dtb-$(CONFIG_ARCH_SOPHGO) += sg2002-nanokvm-pcie.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2042-milkv-pioneer.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2042-evb-v1.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2042-evb-v2.dtb
diff --git a/arch/riscv/boot/dts/sophgo/sg2002-nanokvm-pcie.dts b/arch/riscv/boot/dts/sophgo/sg2002-nanokvm-pcie.dts
new file mode 100644
index 000000000..b3be2b7bb
--- /dev/null
+++ b/arch/riscv/boot/dts/sophgo/sg2002-nanokvm-pcie.dts
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright (C) 2026 Sodo
+ */
+
+/dts-v1/;
+
+#include "sg2002.dtsi"
+
+/ {
+	model = "Sipeed NanoKVM PCIe";
+	compatible = "sipeed,nanokvm-pcie", "sophgo,sg2002";
+
+	aliases {
+		gpio0 = &gpio0;
+		gpio1 = &gpio1;
+		gpio2 = &gpio2;
+		gpio3 = &gpio3;
+		serial0 = &uart0;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+};
+
+&osc {
+	clock-frequency = <25000000>;
+};
+
+&pinctrl {
+	uart0_cfg: uart0-cfg {
+		uart0-pins {
+			pinmux = <PINMUX(PIN_UART0_TX, 0)>,
+				 <PINMUX(PIN_UART0_RX, 0)>;
+			bias-pull-up;
+			drive-strength-microamp = <10800>;
+			power-source = <3300>;
+		};
+	};
+
+	sdhci0_cfg: sdhci0-cfg {
+		sdhci0-clk-pins {
+			pinmux = <PINMUX(PIN_SD0_CLK, 0)>;
+			bias-pull-up;
+			drive-strength-microamp = <16100>;
+			power-source = <3300>;
+		};
+
+		sdhci0-cmd-pins {
+			pinmux = <PINMUX(PIN_SD0_CMD, 0)>;
+			bias-pull-up;
+			drive-strength-microamp = <10800>;
+			power-source = <3300>;
+		};
+
+		sdhci0-data-pins {
+			pinmux = <PINMUX(PIN_SD0_D0, 0)>,
+				 <PINMUX(PIN_SD0_D1, 0)>,
+				 <PINMUX(PIN_SD0_D2, 0)>,
+				 <PINMUX(PIN_SD0_D3, 0)>;
+			bias-pull-up;
+			drive-strength-microamp = <10800>;
+			power-source = <3300>;
+		};
+	};
+};
+
+&sdhci0 {
+	pinctrl-0 = <&sdhci0_cfg>;
+	pinctrl-names = "default";
+	bus-width = <4>;
+	disable-wp;
+	no-1-8-v;
+	no-mmc;
+	no-sdio;
+	status = "okay";
+};
+
+&uart0 {
+	pinctrl-0 = <&uart0_cfg>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&i2c4 {
+	clock-frequency = <100000>;
+	status = "okay";
+
+	hdmi-bridge@2b {
+		compatible = "lontium,lt6911uxe";
+		reg = <0x2b>;
+
+		port {
+			lt6911_out: endpoint {
+				remote-endpoint = <&csi_in>;
+				clock-lanes = <2>;
+				data-lanes = <0 1 3 4>;
+				lane-polarities = <1 1 1 1 1>;
+				link-frequencies = /bits/ 64 <600000000>;
+			};
+		};
+	};
+};
+
+&usb {
+	dr_mode = "peripheral";
+	status = "okay";
+};
+
+&{/soc} {
+	csi: csi@a0c2000 {
+		compatible = "sophgo,sg2002-csi";
+		reg = <0x0a0c2000 0x1000>,
+		      <0x0a0d0000 0x1000>,
+		      <0x0a000000 0x80000>;
+		reg-names = "csi-mac", "csi-wrap", "vi";
+		clocks = <&clk CLK_AXI_VIP>,
+			 <&clk CLK_CSI_MAC0_VIP>,
+			 <&clk CLK_CSI0_RX_VIP>,
+			 <&clk CLK_CFG_REG_VIP>,
+			 <&clk CLK_CAM0_VIP>;
+		resets = <&rst RST_VIPSYS>, <&rst RST_VIP_CAM0>;
+		status = "okay";
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			port@0 {
+				reg = <0>;
+
+				csi_in: endpoint {
+					remote-endpoint = <&lt6911_out>;
+					clock-lanes = <2>;
+					data-lanes = <0 1 3 4>;
+					lane-polarities = <1 1 1 1 1>;
+				};
+			};
+		};
+	};
+};
